File: famous-algorithms/topological-sort/javascript/KahnsAlgorithm.js
class Graph{  
    
    constructor(numberOfVertices) {
        this.numberOfVertices = numberOfVertices;
        this.adjacencyList = [];
        
        for (let i = 0; i < this.numberOfVertices; i++) {
            this.adjacencyList[i] = []
        }
    }
    
    addEdge(u, v) {
        this.adjacencyList[u].push(v);  
    }

    topologicalSort() {
        
        const indegree = new Array(this.numberOfVertices).fill(0);
        
        for (let i = 0; i < this.numberOfVertices; i++) {

            console.log(this.adjacencyList);
            
            //indegree
            
            this.adjacencyList[i].forEach(vertex =>indegree[vertex]=indegree[vertex]+1);
            
        }
        
        //the vertex that we are selecting
        
        const queue = [];
        
        for (let i = 0; i < this.numberOfVertices; i++) {
            //console.log(indegree);
            if(indegree[i] == 0)
            queue.unshift(i);
            console.log(queue)
        }
        
        //keep track of the number selected vertices
        let c = 0;
        
        const linearOrder = [];
        
        while(queue.length !== 0) {
            const x = queue.shift();
            linearOrder.push(x);
            
            this.adjacencyList[x].forEach(vertex => {
                if(--indegree[vertex] == 0)
                queue.unshift(vertex);
            });

            c++;
        }
        
        if(c != this.numberOfVertices) {
            console.log(c + "="+ this.numberOfVertices)
            console.log("Graph contains -ve cycle.");
            return;
        }
        
        console.log(" The Linear Order is :");
        
        linearOrder.forEach(e => console.log(e + " "));
        
    }

}


const graph = new Graph(5);
graph.addEdge(0, 1);
graph.addEdge(0, 3);
graph.addEdge(1, 2);
graph.addEdge(3, 1);
graph.addEdge(3, 2);
graph.addEdge(3, 4);

graph.topologicalSort();

File: famous-algorithms/topological-sort/java/KahnsAlgorithm.java
package com.basicsstrong.algorithm.famousalgorithm;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.Vector;

public class KahnsAlgorithm {
	
	static class Graph{
		
		int V;
		List<Integer>[] adj;
		
		public Graph(int V) {
			this.V = V;
			adj = new ArrayList[V];
			
			for (int i = 0; i < adj.length; i++) {
				adj[i] = new ArrayList<Integer>();
			}
		}
		
		public void addEdge(int u, int v) {
			adj[u].add(v);  
		}

		public void topologicalSort() {
			
			int[] indegree = new int[V];
			
			for (int i = 0; i < V; i++) {
				ArrayList<Integer> temp = (ArrayList<Integer>) adj[i];
				
				//indegree
				for (int vertex : temp) {
					
					indegree[vertex]++;
					
				}
			}
			
			//the vertex that we are selecting
			
			Queue<Integer> q = new LinkedList<Integer>();
			
			for (int i = 0; i < V; i++) {
				if(indegree[i] == 0)
					q.add(i);
			}
			
			//keep track of the number selected vertices
			int c = 0;
			
			Vector<Integer> linearOrder = new Vector<Integer>();
			
			while(!q.isEmpty()) {
				int x = q.poll();
				linearOrder.add(x);
				
				for (Integer vertex : adj[x]) {
					if(--indegree[vertex] == 0)
						q.add(vertex);
				}
				c++;
			}
			
			if(c != V) {
				System.out.println("Graph contains -ve cycle.");
				return;
			}
			
			System.out.println(" The Linear Order is :");
			for (int i : linearOrder) {
				
				System.out.println(i + " ");
				
			}
			
		}
		
	}
	
	public static void main(String[] args) {
		
		Graph g = new Graph(5);
		g.addEdge(0, 1);
		g.addEdge(0, 3);
		g.addEdge(1, 2);
		g.addEdge(3, 1);
		g.addEdge(3, 2);
		g.addEdge(3, 4);
		
		g.topologicalSort();
		
		//1,4,2,5,3
		
	}
	
	
}



